lib/fluent/plugin/out_splunk_ingest_api.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require 'fluent/plugin/out_splunk'
+require 'openid_connect'
+require 'rack/oauth2'
+require 'multi_json'
+
+module Fluent::Plugin
+  class SplunkIngestApiOutput < SplunkOutput
+    Fluent::Plugin.register_output('splunk_ingest_api', self)
+
+    desc 'Service Client Identifier'
+    config_param :service_client_identifier, :string, default: nil
+
+    desc 'Service Client Secret Key'
+    config_param :service_client_secret_key, :string, default: nil
+
+    desc 'Token Endpoint'
+    config_param :token_endpoint, :string, default: '/system/identity/v1/token'
+
+    desc 'Ingest Api Hostname'
+    config_param :ingest_api_host, :string, default: 'api.splunkbeta.com'
+
+    desc 'Ingest API Tenant Name'
+    config_param :ingest_api_tenant, :string
+
+    desc 'Ingest API Events Endpoint'
+    config_param :ingest_api_events_endpoint, :string, default: '/ingest/v1beta2/events'
+
+    desc 'Debug the HTTP transport'
+    config_param :debug_http, :bool, default: false
+
+    def prefer_buffer_processing
+      true
+    end
+
+    def configure(conf)
+      super
+    end
+
+    def construct_api
+      uri = "https://#{@ingest_api_host}/#{@ingest_api_tenant}#{@ingest_api_events_endpoint}"
+      URI(uri)
+    rescue StandardError
+      raise Fluent::ConfigError, "URI #{uri} is invalid"
+    end
+
+    def format(tag, time, record)
+      format_event(tag, time, record)
+    end
+
+    def format_event(tag, time, record)
+      event = prepare_event_payload(tag, time, record)
+      # Unsure how to drop a record. So append the empty string
+      if event[:body].nil? || event[:body].strip.empty?
+        ''
+      else
+        MultiJson.dump(event) + ','
+      end
+    end
+
+    def prepare_event_payload(tag, time, record)
+      payload = super(tag, time, record)
+      payload[:attributes] = payload.delete(:fields) || {}
```

Comment:
If people knowing only the Splunk Ingest API start to use this plugin it maybe will be confusing if they need to use the fields param to set what in Ingest API is called attributes.